src/override/memcpy.cc

```diff
@@ -0,0 +1,233 @@
+   * off. It defaults to true for debug builds, false for release builds.
+   */
+  static constexpr bool FailFast =
+#ifdef SNMALLOC_FAIL_FAST
+    SNMALLOC_FAIL_FAST
+#else
+#  ifdef NDEBUG
+    true
+#  else
+    false
+#  endif
+#endif
+    ;
+
+  /**
+   * The largest register size that we can use for loads and stores.  These
+   * types are expected to work for overlapping copies: we can always load them
+   * into a register and store them.  Note that this is at the C abstract
+   * machine level: the compiler may spill temporaries to the stack, just not
+   * to the source or destination object.
+   */
+  static constexpr size_t LargestRegisterSize =
+#ifdef __AVX__
+    32
+#elif defined(__SSE__)
+    16
+#else
+    sizeof(uint64_t)
+#endif
+    ;
+
+  /**
+   * Copy a single element of a specified size.  Uses a compiler builtin that
+   * expands to a single load and store.
+   */
+  template<size_t Size>
+  SNMALLOC_FAST_PATH inline void copy_one(void* dst, const void* src)
+  {
+#if __has_builtin(__builtin_memcpy_inline)
+    __builtin_memcpy_inline(dst, src, Size);
+#else
+    // Define a structure of size `Size` that has alignment 1 and a default
+    // copy-assignment operator.  We can then copy the data as this type.  The
+    // compiler knows the exact width and so will generate the correct wide
+    // instruction for us (clang 10 and gcc 12 both generate movups for the
+    // 16-byte version of this when targeting SSE.
+    struct Block
+    {
+      char data[Size];
+    };
+    auto* d = static_cast<Block*>(dst);
+    auto* s = static_cast<const Block*>(src);
+    *d = *s;
+#endif
+  }
+
+  SNMALLOC_SLOW_PATH SNMALLOC_UNUSED_FUNCTION void crashWithMessage
+    [[noreturn]] (
+      void* p, size_t len, const char* msg, decltype(ThreadAlloc::get())& alloc)
+  {
+    // We're going to crash the program now, but try to avoid heap
+    // allocations if possible, since the heap may be in an undefined
+    // state.
+    std::array<char, 1024> buffer;
+    snprintf_l(
+      buffer.data(),
+      buffer.size(),
+      /* Force C locale */ nullptr,
+      "%s: %p is in allocation %p--%p, offset 0x%zx is past the end.\n",
+      msg,
+      p,
+      alloc.template external_pointer<Start>(p),
+      alloc.template external_pointer<OnePastEnd>(p),
+      len);
+    Pal::error(buffer.data());
+  }
+
+  /**
+   * Check whether a pointer + length is in the same object as the pointer.
+   * Fail with the error message from the third argument if not.
+   *
+   * The template parameter indicates whether this is a read.  If so, this
+   * function is a no-op when `CheckReads` is false.
+   */
+  template<bool IsRead = false>
+  SNMALLOC_FAST_PATH inline void
+  check_bounds(const void* ptr, size_t len, const char* msg = "")
+  {
+    if constexpr (!IsRead || CheckReads)
+    {
+      auto& alloc = ThreadAlloc::get();
+      void* p = const_cast<void*>(ptr);
+
+      // FIXME: Overflow checking.
```

Comment:
Could you use `TODO` rather than `FIXME`?

Reply:
I'd prefer to keep FIXME for things that are really important.  The lack of overflow checking is actually a security vulnerability here.  To me, TODO implies 'we should probably do this', FIXME means 'this is really a bug and we should fix it'.